qutip/tests/core/test_coefficient.py

```diff
@@ -263,6 +263,9 @@ def test_CoeffOptions():
     pytest.param("t + (0 if not 'something' else 1)", {},
                  lambda t: t + 1, id="branch")
 ])
+@pytest.mark.filterwarnings(
```

Comment:
It's the string `'key'` that cause an issue. It's not smart enough to detect constant that are supposed to be string in the code string. The _right way_ to do it would be to pass to string as an args:
`coefficient("cos(t*dictionary[key])", args={'dictionary': {'key': 1}, 'key':'key'}` work without warning.

I think we should keep the fallback anyway, we don't want to remake a full parsing code and there will always be hedge case that we did not think about.

Reply:
Ah, I see -- we are allowing bare words inside the string.

I would be in favor of deprecating that for version 5 (i.e. to make `t * dictionary["key"]` required if one wants a constant. That's simpler and consistent with Python. However, that's not a discussion for this PR, I don't think.